src/main/java/com/mongodb/jdbc/MongoConnection.java

```diff
@@ -421,38 +421,39 @@ public Void call() throws SQLException {
         }
     }
 
-    @Override
-    public boolean isValid(int timeout) throws SQLException {
+    /**
+     * Executes a dummy query to test the connection.
+     *
+     * @param timeout The query timeout.
+     * @throws Exception If an error occurs.
+     */
+    protected void testConnection(int timeout)
+            throws SQLException, InterruptedException, ExecutionException, TimeoutException {
         if (timeout < 0) {
             throw new SQLException("Input is invalid.");
         }
 
         if (isClosed) {
-            return false;
+            throw new SQLException("Connection is closed.");
         }
         // We use createStatement to test the connection. Since we are not allowed
         // to set the timeout adhoc on the calls, we use Executor to run a blocked call with timeout.
         ExecutorService executor = Executors.newCachedThreadPool();
```

Comment:
This executor is not being shutdown, we can add a `try { } finally { executor.shutdown(); }`

Reply:
oh, yes, added, thank you!